experimental/Schemes/SpecOpen.jl

```diff
@@ -145,15 +156,28 @@ function SpecOpen(X::AbsSpec, I::MPolyIdeal; check::Bool=true)
   return SpecOpen(X, [g for g in gens(I) if !iszero(OO(X)(g))], check=check)
 end
 
+@doc Markdown.doc"""
+  complement(X::Scheme,Y::Scheme) -> Scheme
+
+Return complement `X \ Y` of `Y` in `X`.
+
+Since we want the complement `U = X \ Y` to be a Scheme,
+we require that `Y` is closed in `X`.
+"""
+complement(X::Scheme,Y::Scheme)
+
 function complement(X::AbsSpec, Z::AbsSpec{<:Ring, <:MPolyRing})
+  ambient_ring(X) == ambient_ring(Z) || error("X and Z do not compare")
   return EmptyScheme(base_ring(X))
 end
 
 function complement(X::AbsSpec, Z::AbsSpec{<:Ring, <:MPolyQuo})
+  ambient_ring(X) == ambient_ring(Z) || error("X and Z do not compare")
   return SpecOpen(X, modulus(OO(Z)))
 end
 
 function complement(X::AbsSpec, Z::AbsSpec{<:Ring, <:MPolyLocalizedRing})
+  ambient_ring(X) == ambient_ring(Z) || error("X and Z do not compare")
   return subscheme(X, prod(denominators(inverted_set(OO(Z)))))
```

Comment:
Am I right that this only makes sense for localizations at powers of elements?

Reply:
Yes. And even worse: This is the type of complement that we decided to discard in the above discussion. So it should go away completely.